sound/soc/sof/amd/acp-ipc.c

```diff
@@ -0,0 +1,214 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// This file is provided under a BSD license.  When using or
+// redistributing this file, you may do so under this license.
+//
+// Copyright(c) 2021 Advanced Micro Devices, Inc. All rights reserved.
+//
+// Authors: Balakishore Pati <Balakishore.pati@amd.com>
+//	    Ajit Kumar Pandey	<AjitKumar.Pandey@amd.com>
+
+/* ACP-specific SOF IPC code */
+
+#include <linux/pci.h>
+#include <linux/module.h>
+#include <linux/io.h>
+#include <linux/platform_device.h>
+#include <linux/interrupt.h>
+#include <linux/pm_runtime.h>
+#include <linux/delay.h>
+#include <sound/sof.h>
+#include <sound/sof/xtensa.h>
+#include "../ops.h"
+
+#include "acp.h"
+
+void acpbus_trigger_host_to_dsp_swintr(struct acp_dev_data *adata)
+{
+	struct snd_sof_dev *sdev = adata->dev;
+	u32 swintr_trigger = 0x00;
+
+	swintr_trigger = snd_sof_dsp_read(sdev, ACP_DSP_BAR,
+					  ACP_SW_INTR_TRIG);
+	swintr_trigger |= 0x01;
+	snd_sof_dsp_write(sdev, ACP_DSP_BAR, ACP_SW_INTR_TRIG,
+			  swintr_trigger);
+}
+
+void sof_ipc_host_ack_set(struct AcpScratchRegisterConfig *scratch_buf_info)
+{
+	writel(1, (u32 *)(&scratch_buf_info->sofHostAckwrite));
+}
+
+void sof_ipc_host_msg_set(struct AcpScratchRegisterConfig *scratch_buf_info)
+{
+	writel(1, (u32 *)(&scratch_buf_info->sofHostMsgWrite));
+}
+
+static void acp_dsp_ipc_host_done(struct AcpScratchRegisterConfig *scratch_buf_info)
+{
+	 writel(0, (u32 *)(&scratch_buf_info->sofDspMsgWrite));
+}
+
+static void acp_dsp_ipc_dsp_done(struct AcpScratchRegisterConfig *scratch_buf_info)
+{
+	 writel(0, (u32 *)(&scratch_buf_info->sofDspAckwrite));
+}
+
+/*
+ * This function is called when host has some thing to send
+ * data to DSP.So the direction is Host->DSP
+ * @sdev: sof  device.
+ * @msg : Which has the msg data to send to DSP.
+ *
+ */
+int acp_sof_ipc_send_msg(struct snd_sof_dev *sdev, struct snd_sof_ipc_msg *msg)
+{
+	struct acp_dev_data *adata = sdev->pdata->hw_pdata;
+	unsigned char *ring_buffer = adata->acp_sregmap->scratch_buf_info.sofInBoxBuffer;
+	int HostMsgWrite = 0;
+
+	acp_copy_to_scratch_memory(sdev, ring_buffer, 0, msg->msg_data,
+				   msg->msg_size);
+	sof_ipc_host_msg_set(&adata->acp_sregmap->scratch_buf_info);
+	HostMsgWrite = readl(&adata->acp_sregmap->scratch_buf_info.sofHostMsgWrite);
+
+	/* Trigger host to dsp interrupt for the msg */
+	acpbus_trigger_host_to_dsp_swintr(adata);
+	return 0;
+}
+EXPORT_SYMBOL_NS(acp_sof_ipc_send_msg, SND_SOC_SOF_AMD_COMMON);
+
+void acp_dsp_ipc_get_reply(struct snd_sof_dev *sdev)
+{
+	struct snd_sof_ipc_msg *msg = sdev->msg;
+	struct sof_ipc_reply reply;
+	struct sof_ipc_cmd_hdr *hdr;
+	int ret = 0;
```

Comment:
unnecessary init. infact i think even the variable is not necessary. Why not just use msg->reply_error directly?